Add Lutron fan platform

dcode opened the pull request:
Revisiting an old PR (#43915) that I didn't get around to finishing.

<!--
- Adds LutronFan class
- Moved fan controller into separate file
- Fixed black formatting
- Fix typos
- Remove asyncio, which was incorrect
- Fix import sorts
- First lines end with periods
- change string to comment
- Fix discrete speed logic
- Fix formatting
- Fix isort
- Updated to use new Fan API
-->

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Previously, Lutron Fans were treated like any other dimmer. Now the fan objects will be correctly recognized as ceiling fans with discrete speeds. If user automations previously relied on the dimmer functionality, those automations _might_ break. The percentage value would still work via the new API.

This also adds a `unique_id` to the LutronDevice object.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This change creates the discrete states for a four-speed fan controller and fan entity type.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

dcode committed:
Adds LutronFan class

This is a first, untested stab based upon my understanding of pylutron and interaction with devices using ipython + pylutron. I'm not certain about how I'm registering the devices to hass.

dcode committed:
Moved fan controller into separate file

- Removed MEDIUM_HIGH speed until architectural adjustments are made

dcode committed:
Fixed black formatting

dcode committed:
Fix typos

dcode committed:
Remove asyncio, which was incorrect

dcode committed:
Fix import sorts

dcode committed:
First lines end with periods

dcode committed:
change string to comment

dcode committed:
Fix discrete speed logic

dcode committed:
Fix formatting

dcode committed:
Fix isort

dcode committed:
Updated to use new Fan API

dcode committed:
Formatted and isorted

dcode commented:
I'm not sure if the `version` field is needed in the manifest.json. I had to add it to test as a custom component.

JonGilmore commented:
Hm, my dev environment hardware recently took a dump so I'm trying to test this just within the devcontainer on my macbook, but am facing an issue, I'm getting this error after I setup lutron in the config: 

```
2021-06-24 02:13:56 ERROR (MainThread) [homeassistant.setup] Error during setup of component lutron
Traceback (most recent call last):
  File "/workspaces/core/homeassistant/setup.py", line 249, in _async_setup_component
    result = await task
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/workspaces/core/homeassistant/components/lutron/__init__.py", line 59, in setup
    hass.data[LUTRON_CONTROLLER].load_xml_db()
  File "/usr/local/lib/python3.9/site-packages/pylutron/__init__.py", line 534, in load_xml_db
    assert(parser.parse())     # throw our own exception
  File "/usr/local/lib/python3.9/site-packages/pylutron/__init__.py", line 247, in parse
    for group_xml in groups.getiterator('OccupancyGroup'):
AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getiterator'
```

Trying to figure out if this is a local machine issue or if its because of your changes.

JonGilmore commented:
@dcode would you mind rebasing? This will pull in the latest pylutron from pypi which now includes the fix for the error I was seeing when testing :D

dcode commented:
oh, I merged in the wrong branch. I'll fix it

dcode committed:
Fix for unique_id for buttons without ID attribute

bdraco commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,89 @@
+
+    def set_percentage(self, percentage: int) -> None:
+        """Set the speed of the fan."""
+
+        if percentage == 0:
+            named_speed = FAN_OFF
+        else:
+            named_speed = percentage_to_ordered_list_item(
+                ORDERED_NAMED_FAN_SPEEDS, percentage
+            )
+
```
```suggestion
```

bdraco commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,89 @@
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {"lutron_integration_id": self._lutron_device.id}
+
+    @property
+    def speed_count(self) -> int:
+        """Return the number of speeds the fan supports."""
+        return len(ORDERED_NAMED_FAN_SPEEDS)
```
```suggestion
```

The default is 100

bdraco commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,89 @@
+FAN_OFF, FAN_LOW, FAN_MEDIUM, FAN_MEDIUM_HIGH, FAN_HIGH = 0, 25, 50, 75, 100
+
+DEFAULT_ON_PERCENTAGE = 50
+ORDERED_NAMED_FAN_SPEEDS = [FAN_LOW, FAN_MEDIUM, FAN_MEDIUM_HIGH, FAN_HIGH]
```
```suggestion
```

bdraco commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,89 @@
+
+from . import DOMAIN as LUTRON_DOMAIN, LUTRON_CONTROLLER, LUTRON_DEVICES, LutronDevice
+
+FAN_OFF, FAN_LOW, FAN_MEDIUM, FAN_MEDIUM_HIGH, FAN_HIGH = 0, 25, 50, 75, 100
```
```suggestion
```

bdraco commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,89 @@
+"""Support for Lutron fans."""
+
+from homeassistant.components.fan import DOMAIN, SUPPORT_SET_SPEED, FanEntity
+from homeassistant.util.percentage import (
+    ordered_list_item_to_percentage,
+    percentage_to_ordered_list_item,
+)
+
```
```suggestion
```

bdraco commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,89 @@
+
+    def turn_off(self, **kwargs) -> None:
+        """Turn the fan off."""
+        self.set_percentage(0)
```
```suggestion
        self._lutron_device.level = 0
```

bdraco commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,89 @@
+        super().__init__(area_name, lutron_device, controller)
+
+    @property
+    def device_state_attributes(self):
```
```suggestion
    def extra_state_attributes(self):
```

bdraco commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,89 @@
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {"lutron_integration_id": self._lutron_device.id}
```
Ideally this is provided by `LutronDevice`

dcode replied:
Good point, was just trying to isolate my changes, but I'll move it there.

bdraco reviewed (changes requested):
Since the fan works natively in percentage, we don't need to implement named speeds.  The fan integration will take care of converting back to the legacy names

cdheiser commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -128,6 +131,11 @@ def should_poll(self):
+    @property
+    def unique_id(self):
+        """Return the unique ID of the device."""
+        return f"{DOMAIN}_{self._controller.guid}_{ getattr(self._lutron_device, 'uuid', '')}_{getattr(self._lutron_device, 'id', '')}"
```
Why include the integration ID as well?

dcode replied:
It's a fair point. I don't actually understand how Lutron uses the UUID sufficiently. Unfortunately, not everything in RadioRA2 has a serial number that's exposed, and the official guidance is to use the integration ID. I'll remove it locally and see how it works with my setup. I have a variety of devices, but probably not everything.

dcode commented:
> Since the fan works natively in percentage, we don't need to implement named speeds. The fan integration will take care of converting back to the legacy names

Didn't realize that. I'll remove the named speeds.

dcode committed:
Removed named speeds

Named speeds are handled in HA Fan device. No longer needed here since percentages are native.

Co-authored-by: J. Nick Koston <nick@koston.org>

dipeshc commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,69 @@
+
+    def __init__(self, area_name, lutron_device, controller):
+        """Initialize the fan device."""
+        self._prev_speed = None
```
Looks like `_prev_speed` is set here but then never used. Suspect this is an artifact of copy/pasta from https://github.com/home-assistant/core/blob/dev/homeassistant/components/lutron/light.py#L36.

Rather than cleaning it up, I would actually recommend, sticking with this variable and using it below in the the `turn_on(...)` method where if a percentage is not provided, it will set the previous speed.  From an end user experience, turning off then on the fan via HA, will not cause the fan to lose its previous speed and go back to the `DEFAULT_ON_PERCENTAGE`.

For reference I had actually implemented a duplicate [PR](https://github.com/home-assistant/core/pull/56959/files#diff-80e24b982558092127a33e624eb8654a54980052f8b35a84ec5de989892a1e7fR22) using this approach as it aligned with how the dimmer/light entity has been implemented.

dipeshc commented on homeassistant/components/lutron/fan.py:
```diff
@@ -0,0 +1,69 @@
+    add_entities(devs, True)
+
+
+class LutronFan(LutronDevice, FanEntity):
```
Probably make sense to override `speed_count` to 4. Otherwise by default it will only have 3 which wont match the RRD-2ANF.

bdraco commented:
Looks like this has a conflict.

Also, please request a review on the next turn as I won't see if again unless you do. 👍

frenck commented:
Because there hasn't been any activity on this PR for quite some time now, I've decided to close it for being stale.

Feel free to re-open this PR when you are ready to pick up work on it again 👍
